### PR TITLE
Check that doctrine entities are not final

### PR DESCRIPTION
**Pull request by enumag**
I create all my classes as final by default. This however can cause issues with Doctrine:

```
  [Doctrine\Common\Proxy\Exception\InvalidArgumentException]
  Unable to create a proxy for a final class "<entity class>".
```

What do you think about a rule to avoid final entities?

### CODE REVIEW

File: src/Rules/Doctrine/ORM/EntityRule.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+			return [];
+		}
+
+		$classMetadataInfo = 'Doctrine\ORM\Mapping\ClassMetadataInfo';
+		if (!$metadata instanceof $classMetadataInfo) {
+			return [];
+		}
+
+		$errors = [];
+		if ($node->isFinal()) {
```
Comment (ondrejmirtes) [inline src/Rules/Doctrine/ORM/EntityRule.php:58]: Ask this first, and only if it's final then continue with checking whether it's an entity etc.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Yeah, I'm fine with this. Please add it as a bleeding edge rule similar to here: https://github.com/phpstan/phpstan-strict-rules/blob/334898a32217e4605e0f9cfa3d3fc3101bda26be/rules.neon#L67-L69

**Comment by enumag** [comment]
@ondrejmirtes Updated according to your suggestions. Also added a test case.

**Comment by enumag** [comment]
Hmm... there is already a similar pull request: https://github.com/phpstan/phpstan-doctrine/pull/84

Except I slightly disagree with the statement that entities shouldn't have final methods. I have a case where I'm actually using final method on purpose. But then again I could just disable that particular rule in my codebase.

File: rules.neon [hunk 2]
```diff
@@ -33,3 +33,7 @@ services:
 			reportUnknownTypes: %doctrine.reportUnknownTypes%
 		tags:
 			- phpstan.rules.rule
+
+conditionalTags:
+	PHPStan\Rules\Doctrine\ORM\EntityRule:
```
Comment (ondrejmirtes) [inline rules.neon:38]: Please rename the rule to something like `FinalEntityClassRule` and also add it to `services` (without any tags).

I also guess this is the reason why the build failed, so please fix that.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
I'm fine with the rule in this PR as it is.

**Comment by enumag** [comment]
Done. :-)

File: rules.neon [hunk 3]
```diff
@@ -19,6 +19,7 @@ rules:
 	- PHPStan\Rules\Doctrine\ORM\DqlRule
 	- PHPStan\Rules\Doctrine\ORM\RepositoryMethodCallRule
 	- PHPStan\Rules\Doctrine\ORM\EntityRelationRule
+	- PHPStan\Rules\Doctrine\ORM\EntityNotFinalRule
```
Comment (ondrejmirtes) [inline rules.neon:22]: It has to be added to `services`. This way you're always registering the rule and the `conditionalTags` section has no effect.

**Comment by enumag** [comment]
Should be okay now.

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by ondrejmirtes** [comment]
Tagged as 0.12.20.

**Comment by mhujer** [comment]
@enumag thanks! 👍

**Pull request merged**